### ISSUE TITLE
Bug in Sys::Syslog in perl 5.8.0

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6252]
Migrated from [rt.perl.org#20557](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=20557) (status was 'resolved')

Searchable as RT20557$

**Comment by p5pRT** [comment]
### From gahan@pschmidt.org


If this is the wrong place for this report I apologize\.  Please point me
at the correct destination and I will forward this accordingly\.

In the Sys&#8203;::Syslog&#8203;::syslog that's bundled with 5\.8\.0 around line 277
there's a spurious or more likely\, a misplaced "&connect unless
$connnected;"\.  Since it appears before the sub is finished parsing all
it's parameters it breaks the %m/$\! functionality by resetting $\!\.

So it either needs to be moved after the line&#8203;:

&nbsp;    $message = sprintf\($mask\, @&#8203;\_\);

or

local\($errno\) = $\!;

needs to be added at the beginnning of the function and

&nbsp;    $mask =~ s/%m/$\!/g;

needs to be changed to

&nbsp;    $mask =~ s/%m/$errno/g;

Thanx for all the work you all put into making perl the treasure it
is\.  Hope this helps\.

Cheers\,
Paul R\. Schmidt
gahan@&#8203;pschmidt\.org

**Comment by p5pRT** [comment]
### From @floatingatoll

\[gahan@&#8203;pschmidt\.org \- Sun Jan 26 13&#8203;:45&#8203;:24 2003\]&#8203;:

> If this is the wrong place for this report I apologize\.  Please point me
> at the correct destination and I will forward this accordingly\.

You've got the right place :\)

> In the Sys&#8203;::Syslog&#8203;::syslog that's bundled with 5\.8\.0 around line 277
> there's a spurious or more likely\, a misplaced "&connect unless
> $connnected;"\.  Since it appears before the sub is finished parsing all
> it's parameters it breaks the %m/$\! functionality by resetting $\!\.

Here's a patch against the development branch of Perl\, that applies this fix; if it's accepted\, it'll be in the next release of Perl\.

Thanks\!

 \- R\.

\# attached\, in case of whitespace mangling
<details><summary>Inline Patch</summary>

```diff
--- ext/Sys/Syslog/Syslog.pm.orig       2003-02-15 01:38:53.000000000 -0500
+++ ext/Sys/Syslog/Syslog.pm    2003-02-15 01:39:48.000000000 -0500
@@ -230,6 +230,7 @@
 }
 
 sub syslog {
+    local($errno) = shift;
     local($priority) = shift;
     local($mask) = shift;
     local($message, $whoami);
@@ -282,7 +283,7 @@
 
     $whoami .= "[$$]" if $lo_pid;
 
-    $mask =~ s/%m/$!/g;
+    $mask =~ s/%m/$errno/g;
     $mask .= "\n" unless $mask =~ /\n$/;
     $message = sprintf ($mask, @_);
 
```
</details>

**Comment by p5pRT** [comment]
### From @floatingatoll

<details><summary>20557-save_errno.patch</summary>

```diff
--- ext/Sys/Syslog/Syslog.pm.orig	2003-02-15 01:38:53.000000000 -0500
+++ ext/Sys/Syslog/Syslog.pm	2003-02-15 01:39:48.000000000 -0500
@@ -230,6 +230,7 @@
 }
 
 sub syslog {
+    local($errno) = shift;
     local($priority) = shift;
     local($mask) = shift;
     local($message, $whoami);
@@ -282,7 +283,7 @@
 
     $whoami .= "[$$]" if $lo_pid;
 
-    $mask =~ s/%m/$!/g;
+    $mask =~ s/%m/$errno/g;
     $mask .= "\n" unless $mask =~ /\n$/;
     $message = sprintf ($mask, @_);
 

```
</details>

**Comment by p5pRT** [comment]
@floatingatoll - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From gahan@pschmidt.org


This fix will break the whole interface to the syslog function for
anything that currently uses it\.  Also if the caller needs to
explicitly pass the errno into the function then they're better off
doing it manually via the printf style format and args than putting it
at the head of the parm list\.

Your fix would work fine if you changed the&#8203;:

&nbsp;    local\($errno\) = shift;

to&#8203;:

&nbsp;    local\($errno\) = $\!;

or to avoid adding any additional code just move the first "&connect
unless $connnected;" after the block of code that follows it \(where the
value of errno is substituted into the format string"\.  I guess I don't
understand why the "&connect unless $connnected;" that is breaking the
defined functionality of the routine needs to be buried in the middle
of the code to parse options\.

\-\- 

Cheers\,
Paul R\. Schmidt
gahan@&#8203;pschmidt\.org

On 15 Feb 2003\, crystalflame wrote&#8203;:

> Date&#8203;: 15 Feb 2003 06&#8203;:42&#8203;:53 \-0000
> From&#8203;: crystalflame \<perlbug\-followup@&#8203;perl\.org>
> To&#8203;: gahan@&#8203;pschmidt\.org
> Subject&#8203;: \[perl \#20557\] Bug in Sys&#8203;::Syslog in perl 5\.8\.0
> 
> \[gahan@&#8203;pschmidt\.org \- Sun Jan 26 13&#8203;:45&#8203;:24 2003\]&#8203;:
> 
> > If this is the wrong place for this report I apologize\.  Please point me
> > at the correct destination and I will forward this accordingly\.
> 
> You've got the right place :\)
> 
> > In the Sys&#8203;::Syslog&#8203;::syslog that's bundled with 5\.8\.0 around line 277
> > there's a spurious or more likely\, a misplaced "&connect unless
> > $connnected;"\.  Since it appears before the sub is finished parsing all
> > it's parameters it breaks the %m/$\! functionality by resetting $\!\.
> 
> Here's a patch against the development branch of Perl\, that applies this fix; if it's accepted\, it'll be in the next release of Perl\.
> 
> Thanks\!
> 
>  \- R\.
> 
> \# attached\, in case of whitespace mangling
> \-\-\- ext/Sys/Syslog/Syslog\.pm\.orig       2003\-02\-15 01&#8203;:38&#8203;:53\.000000000 \-0500
> \+\+\+ ext/Sys/Syslog/Syslog\.pm    2003\-02\-15 01&#8203;:39&#8203;:48\.000000000 \-0500
> @&#8203;@&#8203; \-230\,6 \+230\,7 @&#8203;@&#8203;
>  \}
>  
>  sub syslog \{
> \+    local\($errno\) = shift;
>      local\($priority\) = shift;
>      local\($mask\) = shift;
>      local\($message\, $whoami\);
> @&#8203;@&#8203; \-282\,7 \+283\,7 @&#8203;@&#8203;
>  
>      $whoami \.= "\[$$\]" if $lo\_pid;
>  
> \-    $mask =~ s/%m/$\!/g;
> \+    $mask =~ s/%m/$errno/g;
>      $mask \.= "\\n" unless $mask =~ /\\n$/;
>      $message = sprintf \($mask\, @&#8203;\_\);
>

**Comment by p5pRT** [comment]
### From gahan@pschmidt.org

<details><summary>20557-save_errno.patch</summary>

```diff
--- ext/Sys/Syslog/Syslog.pm.orig	2003-02-15 01:38:53.000000000 -0500
+++ ext/Sys/Syslog/Syslog.pm	2003-02-15 01:39:48.000000000 -0500
@@ -230,6 +230,7 @@
 }
 
 sub syslog {
+    local($errno) = shift;
     local($priority) = shift;
     local($mask) = shift;
     local($message, $whoami);
@@ -282,7 +283,7 @@
 
     $whoami .= "[$$]" if $lo_pid;
 
-    $mask =~ s/%m/$!/g;
+    $mask =~ s/%m/$errno/g;
     $mask .= "\n" unless $mask =~ /\n$/;
     $message = sprintf ($mask, @_);
 

```
</details>

**Comment by p5pRT** [comment]
### From @maddingue

The next release of Sys&#8203;::Syslog will include the modification to save 
errno before trying to connect\. Thanks for your report\. 

\-\- 
Close the world\, txEn eht nepO\.

**Comment by p5pRT** [comment]
### From @maddingue

This is fixed in Sys&#8203;::Syslog 0\.16\, now on the CPAN\. 

\-\- 
Close the world\, txEn eht nepO\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
